### ISSUE TITLE
Question: Anyone to share any documented example please?

### ISSUE THREAD

**Issue by stefanos82** [krakjoe/parallel#34]
I'm using Golang whenever I need concurrency and parallelism, but I would like to stick with PHP for achieving either the same results or the 85% the least of actual performance if possible.

Is there any project / example documented somewhere so I can read about it and teach myself how to use his extension extensively?

**Comment by WyriHaximus** [comment]
Anything you'de looking for specifically? I've put up my PoC tests for [`react-parallel`](https://github.com/wyrihaximus/reactphp-parallel) up [here](https://gist.github.com/WyriHaximus/ef8c8d23d477a72470810ee4afd59685). Which includes a basic test, one doing json decoding (you'll need a json file for it tho), and one doing channels (made by jouw).

**Comment by krakjoe** [comment]
The documentation in the manual, code in tests, and the few gists that I and others have posted is all there is right now ...

I intend to do a tutorial section in the manual, based upon a walk through of building [this](https://gist.github.com/krakjoe/0ee02b887288720d9b785c9f947f3a0a), which covers many of the topics and questions someone new to this kind of concurrency might have, unfortunately it's only lightly commented code right now. It will take a day or two to write up manual pages, and I don't know when I'll have the time to do that.

I would think someone familiar with go should be able to get started just by reading the manual, it's not the same, but should be familiar enough, I think.

**Comment by stefanos82** [comment]
> Anything you'de looking for specifically? I've put up my PoC tests for [`react-parallel`](https://github.com/wyrihaximus/reactphp-parallel) up [here](https://gist.github.com/WyriHaximus/ef8c8d23d477a72470810ee4afd59685). Which includes a basic test, one doing json decoding (you'll need a json file for it tho), and one doing channels (made by jouw).

Thank you @WyriHaximus I will have a look :+1:

**Comment by stefanos82** [comment]
> The documentation in the manual, code in tests, and the few gists that I and others have posted is all there is right now ...
> 
> I intend to do a tutorial section in the manual, based upon a walk through of building [this](https://gist.github.com/krakjoe/0ee02b887288720d9b785c9f947f3a0a), which covers many of the topics and questions someone new to this kind of concurrency might have, unfortunately it's only lightly commented code right now. It will take a day or two to write up manual pages, and I don't know when I'll have the time to do that.
> 
> I would think someone familiar with go should be able to get started just by reading the manual, it's not the same, but should be familiar enough, I think.

In theory yes, I *should* be able to pick up rather quickly, but I haven't read https://php.net/parallel extensively to fully understand it.

I guess I need to invest more time in it.

**Comment by ghost** [comment]
Hi,
how to get the thread pid ? (like getmypid() on your pthreads development ;-) )

and I'm ok so you invest more time in the documentation !

good continuation in your development that interests me a lot (I am currently testing your pthreads library)

**Comment by CViniciusSDias** [comment]
I know this ticket is closed but I think this is the best place to ask...

I'm not a Go programmer but I would still like to learn about this extension. The communication between threads is far from clear to me because I've learned about parallel programming in college using C.

I would just like to see some simple examples or documentation explaining more clearly what are channels and events because from php.net/parallel I could understand much.

Thanks in advance.

**Comment by hdvianna** [comment]
Hi,

I am working on an interface for easing the development of parallel tasks in PHP. It is still very inceptive but I assume that is possible to grasp the idea.

And here goes the project link https://github.com/hdvianna/parallel-workerpool

Kind regards,
Henrique

**Comment by CViniciusSDias** [comment]
Thank you for providing an example, @hdvianna 
But from the time I wrote the message untill now I did a quick research on how Go works and I understood the concept of channels.
The thing that still isn't clear to me is the Event part. =/
And it doesn't seem like you used the Event (or Events) class.

**Comment by awderh** [comment]
Hey @CViniciusSDias 

I became familiarized with Events when I wanted to do a non-blocking recieve.

To me, events are a centralized approach to handling channel/future communications. An "event" could be a read/write/close (etc) on a Channel/Future. Events - it appears - is for use in a loop, where you can regularly poll and then "delegate events" (based on their characteristics).

Consider the following: you want to make a server that runs tasks in the background. It is interfaced via HTTP. Suppose it has one worker thread. You'd do something like this:

```php
<?php
$worker = new \parallel\Runtime();
$workerChannel = new \parallel\Channel(); // for communicating with our worker
$worker->run(function($workerChannel) {
    while (true) { // you'd want a sensible exit point here :) 
        $request = $workerChannel->recv();
        $arrayReturn = magicallyHandleRequest($request );
        $workerChannel->send($arrayReturn);
    }
}, [$workerChannel]);


$h = new HTTPServer(); // your own implementation
$h->run(function(Request $req) use ($workerChannel) {
    $taskName = $req['taskName'];
    $workerChannel->send($taskName); // normally, you'd abstract out HTTP stuffs from your core code, kinda like this, as interfaces may change
    $jsonResponse = new JSONResponse($workerChannel->recv());
    return $jsonResponse;
});
```

So, every time you send a request to the server, a message is sent to your worker (which, btw, is waiting - via blocking recv() - for a command). This does work... but, what if you recieve 1000 requests at once and/or your task will take a long time to complete?

1) Your HTTP Server will not be accepting new requests, until it is done with the current one. 
2) Your worker thread is already pinned so, even if you were to handle a new request, you'd end up blocking anyway.

The client may wait a very long time. As such, one might:
1) Immediately send an acknowledgement response for a new task, or provide output for an existing/completed task
2) Have the worker delegate to a pool, while checking regularly for updates from existing tasks - this will require **Events!**

```php
<?php

$worker = new \parallel\Runtime();
$workerChannel = new \parallel\Channel(10); // **Make it buffered this time**
$worker->run(function($workerChannel) {
    $events = new \parallel\Events();
    $events->addChannel($workerChannel);
    $events->setBlocking(false);

    $futureMap = [];

    while (true) {

        if (($event = $events->poll()) != null) {
            // something happened, let's figure out what it was. First, we check the source.
            if ($event->object == $workerChannel) {
                // OK, now we must add our request to the queue, or return an output if we're already processing it
                if ($event->type == \Parallel\Events\Event\Type::Read) {
                    $request = $event->value;
                    if (!isset($futureMap[$request])) {
                        // here, we setup a task to run in parallel
                        $future = \parallel\Run(function($request) {
                            $arrayReturn = magicallyHandleRequest($request);
                            return ['taskName' => $request, 'output' => $arrayReturn];
                        }, [$request]);
                        // and also add it into our events system, so that we can get data from it without blocking
                        $events->addHandler($future);
                        // finally, a way to output that data later on and track it too
                        $futureMap[$request] = ['output' => '', 'future' => $future];
                    }
                    
                    // the http server is still waiting for a response, so we (relatively) quickly send one back
                    $workerChannel->send($futureMap[$request]['output']);
                }
                $events->addChannel($workerChannel); // gets removed, so we put it back in.
            } else if ($event->object instanceof \parallel\Future) {
                if ($event->type == \Parallel\Events\Event\Type::Read) { // our task finished! 
                    $futureResponse = $event->value;
                    $request = $futureResponse['taskName'];
                    $newOutput = $futureResponse['output'];
                    $futureMap[$request]['output'] = $newOutput;
                    // the next time the client asks for the request, they'll have the output they're looking for
                }
            }
        }
    }
}, [$workerChannel]);


$h = new HTTPServer(); // your own implementation
$h->run(function(Request $req) use ($workerChannel) {
    $taskName = $req['taskName'];
    $workerChannel->send($taskName); // normally, you'd abstract out HTTP stuffs from your core code, kinda like this, as interfaces may change
    $jsonResponse = new JSONResponse($workerChannel->recv());
    return $jsonResponse;
});
```
